File: src/main/java/com/test/Application.java
package com.test;

import com.test.input.GenericService;
import com.test.input.GenericServiceCompositeKey;
import com.test.input.ServiceStatus;
import org.apache.flink.api.common.RuntimeExecutionMode;
import org.apache.flink.api.common.functions.AggregateFunction;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.contrib.streaming.state.EmbeddedRocksDBStateBackend;
import org.apache.flink.state.api.*;
import org.apache.flink.state.api.functions.WindowReaderFunction;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.assigners.GlobalWindows;
import org.apache.flink.streaming.api.windowing.windows.GlobalWindow;
import org.apache.flink.util.Collector;

import java.util.List;

public class Application {
    public static void main(String[] args) throws Exception {
        List<GenericService> services = List.of(
                new GenericService("X", ServiceStatus.ON, "FOO"),
                new GenericService("X", ServiceStatus.ON, "BAR")
        );

        StreamExecutionEnvironment flinkEnv = StreamExecutionEnvironment.getExecutionEnvironment();
        flinkEnv.setParallelism(1);
        flinkEnv.setRuntimeMode(RuntimeExecutionMode.BATCH);

        SavepointWriter savepointWriter = SavepointWriter.newSavepoint(flinkEnv, new EmbeddedRocksDBStateBackend(), 30);

        StateBootstrapTransformation<GenericService> transformation = OperatorTransformation.bootstrapWith(flinkEnv.fromCollection(services))
                .keyBy((value) -> new GenericServiceCompositeKey(value.getServiceId(), value.getCountryCode()))
                .window(GlobalWindows.create())
                .aggregate(new AggregateFunctionForMigration());

        savepointWriter.withOperator(OperatorIdentifier.forUid("test"), transformation)
                .write("/tmp/savepoint");

        flinkEnv.execute();
        System.out.println("Savepoint created");

        // -----------------

        SavepointReader.read(flinkEnv, "/tmp/savepoint")
                .window(GlobalWindows.create())
                .aggregate(
                        "test",
                        new AggregateFunctionForMigration(),
                        new StateReader(),
                        TypeInformation.of(GenericServiceCompositeKey.class),
                        TypeInformation.of(GenericService.class),
                        TypeInformation.of(GenericService.class)
                )
                .print();

        flinkEnv.execute();
    }

    private static class AggregateFunctionForMigration implements AggregateFunction<GenericService, GenericService, GenericService> {
        @Override
        public GenericService createAccumulator() {
            return new GenericService();
        }

        @Override
        public GenericService add(GenericService value, GenericService accumulator) {
            return value;
        }

        @Override
        public GenericService getResult(GenericService accumulator) {
            return null;
        }

        @Override
        public GenericService merge(GenericService a, GenericService b) {
            return null;
        }
    }

    private static class StateReader extends WindowReaderFunction<GenericService, GenericService, GenericServiceCompositeKey, GlobalWindow> {
        @Override
        public void readWindow(GenericServiceCompositeKey genericServiceCompositeKey, Context<GlobalWindow> context, Iterable<GenericService> elements, Collector<GenericService> out) {
            System.out.println("KEY=" + genericServiceCompositeKey);
            elements.forEach(e -> {
                if (e == null) {
                    System.out.println("Why is this null?");
                } else {
                    out.collect(e);
                }
            });
        }
    }
}

File: src/main/kotlin/com/test/Application.kt
package com.test

import com.test.input.*
import com.test.state.ActiveServicesCacheBootstrapper
import org.apache.flink.api.common.RuntimeExecutionMode
import org.apache.flink.api.connector.source.Boundedness
import org.apache.flink.contrib.streaming.state.EmbeddedRocksDBStateBackend
import org.apache.flink.state.api.OperatorIdentifier
import org.apache.flink.state.api.OperatorTransformation
import org.apache.flink.state.api.SavepointWriter
import org.apache.flink.streaming.api.datastream.DataStream
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment
import org.apache.flink.streaming.api.functions.ProcessFunction
import org.apache.flink.streaming.api.transformations.LegacySourceTransformation
import org.apache.flink.util.Collector
import java.nio.file.Paths
import kotlin.io.path.ExperimentalPathApi
import kotlin.io.path.deleteRecursively

@OptIn(ExperimentalPathApi::class)
fun main(args: Array<String>) {
    val flinkEnv = StreamExecutionEnvironment.getExecutionEnvironment().apply {
        setRuntimeMode(RuntimeExecutionMode.BATCH)
    }

    val savepointWriter = SavepointWriter.newSavepoint(flinkEnv, EmbeddedRocksDBStateBackend(), 120)

//    val total = args.firstOrNull()?.toIntOrNull() ?: 200_000
//    val inputFormat = TestInputFormat(total)
    val inputFormat = TestNdJsonInputFormat()
    val source = flinkEnv.createInput(inputFormat).forceNonParallel()
    with(source.transformation as LegacySourceTransformation<*>) {
        boundedness = Boundedness.BOUNDED
    }

//    val allowed = setOf(
//        GenericServiceCompositeKey("A"),
//        GenericServiceCompositeKey("B"),
//        GenericServiceCompositeKey("C"),
//        GenericServiceCompositeKey("D"),
//        GenericServiceCompositeKey("E"),
//    )
    val allowed = (1001..1030).mapTo(mutableSetOf()) { GenericServiceCompositeKey(it.toString()) }
    val filtered = source.process(ActiveServicesFilter(allowed))

    configureStateBootstrap(filtered, savepointWriter, "first")
    configureStateBootstrap(filtered, savepointWriter, "second")
    configureStateBootstrap(filtered, savepointWriter, "third")
    configureStateBootstrap(filtered, savepointWriter, "fourth")
    configureStateBootstrap(filtered, savepointWriter, "fifth")
    configureStateBootstrap(filtered, savepointWriter, "sixth")

    savepointWriter.write("file:///tmp/sp")
    flinkEnv.execute()
    println("DONE")
    Paths.get("/tmp/sp").deleteRecursively()
}

private fun configureStateBootstrap(cacheStream: DataStream<Statuses>, savepointWriter: SavepointWriter, uid: String) {
    val transformation = OperatorTransformation
        .bootstrapWith(cacheStream)
        .keyBy { PartitionKey(it.key) }
        .transform(ActiveServicesCacheBootstrapper())

    val id = OperatorIdentifier.forUid(uid)
    savepointWriter
        .removeOperator(id)
        .withOperator(id, transformation)
}

private class ActiveServicesFilter(
    private val allowedServices: Set<GenericServiceCompositeKey>,
) : ProcessFunction<Statuses, Statuses>() {

    companion object {
        private const val serialVersionUID = 1L
    }

    override fun processElement(value: Statuses, ctx: Context, out: Collector<Statuses>) {
        val services = value.services.filter {
            it.status == ServiceStatus.ON &&
                    GenericServiceCompositeKey(it.serviceId, it.countryCode) in allowedServices
        }
        if (services.isNotEmpty()) {
            out.collect(value.copy(services = services))
        }
    }
}
